File: ios/Classes/FlurryPlugin.m
#import "FlurryPlugin.h"
#import "Flurry.h"

@implementation FlurryPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"flurry"
            binaryMessenger:[registrar messenger]];
  FlurryPlugin* instance = [[FlurryPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"initialize" isEqualToString:call.method]) {
        NSDictionary *arguments = (NSDictionary*)call.arguments;
        NSString *apiKey = (NSString*)arguments[@"api_key_ios"];
        BOOL isLogEnabled = (BOOL)arguments[@"is_log_enabled"];

        if (isLogEnabled){

            FlurrySessionBuilder* builder = [[[FlurrySessionBuilder new]
                                                withLogLevel:FlurryLogLevelAll]
                                               withCrashReporting:YES];

            [Flurry startSession:apiKey withSessionBuilder:builder];

        } else {

            FlurrySessionBuilder* builder = [[[FlurrySessionBuilder new]
                                              withLogLevel:FlurryLogLevelNone]
                                             withCrashReporting:YES];

            [Flurry startSession:apiKey withSessionBuilder:builder];
        }

        result(nil);
    } else if ([@"logEvent" isEqualToString:call.method]) {
        NSDictionary *arguments = (NSDictionary*)call.arguments;
        NSString *message = (NSString*)arguments[@"message"];
        [Flurry logEvent:message];
        result(nil);

    } else if ([@"userId" isEqualToString:call.method]) {
        NSDictionary *arguments = (NSDictionary*)call.arguments;
        NSString *userId = (NSString*)arguments[@"userId"];
        [Flurry setUserID:userId];
        result(nil);
    }
    else if ([@"logError" isEqualToString:call.method]) {
            NSDictionary *arguments = (NSDictionary*)call.arguments;
            NSString *errorId = (NSString*)arguments[@"errorId"];
            NSString *exception = (NSString*)arguments[@"exception"];
            NSString *message = (NSString*)arguments[@"message"];

            [Flurry logError:errorId message:message error:exception];
            result(nil);
        }
    else {
        result(FlutterMethodNotImplemented);
    }
}

@end

File: android/src/main/java/hs/flurry/FlurryPlugin.java
package hs.flurry;

import android.content.Context;
import android.util.Log;

import com.flurry.android.FlurryAgent;
import com.flurry.android.FlurryAgentListener;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** FlurryPlugin */
public class FlurryPlugin implements MethodCallHandler {

  private Context activity;
  private MethodChannel channel;

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "flurry");
    channel.setMethodCallHandler(new FlurryPlugin(registrar.activeContext(), channel));
  }

  public FlurryPlugin(Context activity, MethodChannel channel) {
    this.activity = activity;
    this.channel = channel;
    this.channel.setMethodCallHandler(this);
  }

  private void logEvent(String message) {
    FlurryAgent.logEvent(message);
  }

  private void setUserId(String userId) {
    FlurryAgent.setUserId(userId);
  }

  @Override
  public void onMethodCall(MethodCall call, final Result result) {
    if (call.method.equals("initialize")) {
      String API_KEY = call.argument("api_key_android");
      boolean showLog = call.argument("is_log_enabled");

      // initializeFlurry(API_KEY, showLog);

      new FlurryAgent.Builder().withLogEnabled(showLog).withCaptureUncaughtExceptions(true)
          .withContinueSessionMillis(10000).withLogLevel(Log.DEBUG).withListener(new FlurryAgentListener() {
            @Override
            public void onSessionStarted() {
              // result.success(null);
            }
          }).build(activity, API_KEY);

    } else if (call.method.equals("logEvent")) {
      String message = call.argument("message").toString();
      logEvent(message);
      result.success(null);

    } else if (call.method.equals("userId")) {
      String userId = call.argument("userId").toString();
      setUserId(userId);
      result.success(null);

    }
    else if(call.method.equals("logError")){
      String message = call.argument("message").toString();
      String exception = call.argument("exception").toString();
      String errorId = call.hasArgument("errorId")?call.argument("errorId").toString():"";

      FlurryAgent.onError(errorId, message, exception);
      result.success(null);
    }
    else {
      result.notImplemented();
    }
  }
}
